docs/proposals/container-runtime-interface-v1.md

```diff
@@ -0,0 +1,182 @@
+   by [#13768](https://issues.k8s.io/13768), where benefits and shortcomings of
+   such an architecture is discussed.
+ - adding support to Windows containers. Windows container support is a
+   parallel effort and is tracked by [#22623](https://issues.k8s.io/22623). The
+   new interface will leave room for future integration, but will
+ - re-defining Kubelet's internal interfaces. These interfaces, though, may affect
+   Kubelet's maintainability, is not relevant to runtime integration's.
+
+## Requirements
+
+ * Support existing integrated container runtimes
+ * Support hypervisor-based container runtimes
+
+## Container Runtime Interface
+
+The main idea of the proposal is to adopt an imperative container-level
+interface, which allows Kubelet to control the lifecycles of the containers
+inside a pod, across all container runtimes. A separate PodSandbox is concept
+is defined to represent the environment in which a group of containers in a
+pod will be created and run. The container runtimes may interpret the sandbox
+concept differently based on how it operates internally. For runtimes relying
+on hypervisor, sandbox represents a virtual machine. For others, it can be a
+container holding the namespaces.
+
+In short, a PodSandbox should have the following features.
+
+ * **Isolation**: E.g., Linux namespaces or a full virtual machine, or even
+   support additional security features.
+ * **Resource requirements**: A sandbox should implement pod-level resource
+   requirements.
+
+A container in a PodSandbox maps to a application in the Pod Spec. For Linux
+containers, they are expected to share at least network and IPC namespaces,
+with sharing more namespaces discussed in [#1615](https://issues.k8s.io/1615).
+
+
+Below is an example of the proposed interfaces.
+```go
+// PodSandboxManager contains basic operations for sandbox.
+type PodSandboxManager interface {
+	Create(config *PodSandboxConfig) (string, error)
+	Delete(id string) (string, error)
+	List(filter PodSandboxFilter) []PodSandboxListItem
+	Inspect(id string) PodSandbox
+}
+
+// ContainerRuntime contains basic operations for containers.
+type ContainerRuntime interface {
+    Create(config *ContainerConfig, sandboxConfig *PodSandboxConfig, sandboxIDng) (string, error)
+    Start(id string) error
+    Stop(id string, timeout int) error
+    Remove(id string) error
+    List(filter ContainerFilter) ([]ContainerListItem, error)
+    Inspect(id string) (Container, error)
+}
+
+// ContainerCommandExecutor provides methods to run commands in a container.
+type ContainerCommandExecutor interface {
+	Exec(id string, cmd []string, streamOpts StreamOptions) error
+}
+
+// ImageOperations contains image-related operations.
+type ImageOperations interface {
+	List() ([]Image, error)
+	Pull(image ImageSpec, auth AuthConfig) error
+	Remove(image ImageSpec) error
+	Inspect(image ImageSpec) (Image, error)
+}
+
+type PodSandboxMetricsGetter interface {
```

Comment:
Disk usage may be?

On Mon, May 16, 2016 at 4:42 PM, Dawn Chen <notifications@github.com> wrote:

> In docs/proposals/container-runtime-interface-v1.md
> <https://github.com/yujuhong/kubernetes/pull/1#discussion_r63444768>:
>
> > +}
> > +
> > +// ContainerCommandExecutor provides methods to run commands in a container.
> > +type ContainerCommandExecutor interface {
> > +	Exec(id string, cmd []string, streamOpts StreamOptions) error
> > +}
> > +
> > +// ImageOperations contains image-related operations.
> > +type ImageOperations interface {
> > +	List() ([]Image, error)
> > +	Pull(image ImageSpec, auth AuthConfig) error
> > +	Remove(image ImageSpec) error
> > +	Inspect(image ImageSpec) (Image, error)
> > +}
> > +
> > +type PodSandboxMetricsGetter interface {
>
> Are we really expecting each runtime to report the pod level metrics?
> Container-level metrics does make sense, especially for hyper case. But why
> pod-level level? Kubelet has to do resource management at pod level, which
> means kubelet's resource management component has to monitor pod's
> cgroup-level resource usage, even make decision upon out-of-resource
> scenario.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/yujuhong/kubernetes/pull/1/files/ddd041c68ca94d95fa591dce703932fe5d577dd0#r63444768>
>